ipq806x: move common pinmux nodes to SoC DTSI

mans0n opened the pull request:
As almost same pinmux nodes are repeated in each device dts,
let's define them once in the ipq8064 dtsi and remove the rest.

---

For now I only moved `nand_pins` to be safe. My goal is to unify all these pinmux nodes:

* uart0_pins
* i2c4_pins
* spi_pins
* nand_pins
* mdio0_pins
* rgmii2_pins

The other nodes are slightly different for each device dts, I'll add them under your approval.

mans0n committed:
ipq806x: move common pinmux nodes to SoC DTSI

As almost same pinmux nodes are repeated in each device dts,
let's define them once in the ipq8064 dtsi and remove the rest.

* uart0_pins : Did not touch.
* i2c4_pins  : This node seems to be not used at all in dts.
               Added the most common form in the dtsi. Did not touch rpm_pins.
* spi_pins   : The common mux node is already present in the dtsi.
               Removed the duplicate nodes from dts.
* nand_pins  : Moved the entire node.
* mdio0_pins : Moved the common mux node.
* rgmii2_pins: Moved the common mux node. "pins" property is overrided in ap161.

Signed-off-by: Sungbo Eo <mans0n@gorani.run>

adrianschmutzler commented:
I'm asking myself here whether this is the same by coincidence or based on a principle for this particular SoC. In other words: is there a chance we have to move everything back for a specific new device?

adrianschmutzler commented:
@Ansuel Can you comment here?

Ansuel commented:
@adrianschmutzler the pinmux definition is ""standard"" for this soc... in upstream kernel they also have defined some pinmux entires in the dtsi file so I think we can follow the same path.

https://elixir.bootlin.com/linux/latest/source/arch/arm/boot/dts/qcom-ipq8064.dtsi

adrianschmutzler commented:
@mans0n 

From what I checked, it should be trivial to move the nand_pins and mdio0_pins.

For the others, the setups are not entirely equal, so this might be less fun.

Apart from that, one should agree on an order for the nodes in the DTSI.

mans0n commented:
There are two versions of mdio0_pins, one without clk node and one with it. I believe the later is the enhanced version.
I also checked for other pinmuxes.

* uart0_pins: one with 12mA, one with 10mA.
* i2c4_pins: one without drive-strength, one with 10mA, one with 12mA. I think we can just use the third option.
* spi_pins: one with only mux node (which is already in the dtsi), one with two subnodes, one with three subnodes, one with four subnodes, one with four subnodes and gpio cs pin.
* rgmii2_pins: one with only mux node, one with tx node, one with gpio2 and gpio66.

I feel those small differences between devices do not actually matter in most case. That's why I'd like to prevent further variants by unifying the pinmux nodes in the SoC dtsi.

And I don't understand why there are leds_pins and buttons_pins in ipq8064.dtsi, but those are also in upstream kernel so I'm not sure about removing them.

I don't have a strong opinion for the pinmux order. I suggest following the order of function nodes in dtsi.

adrianschmutzler commented:
> There are two versions of mdio0_pins, one without clk node and one with it.

One could move the common part to the DTSI, and then use the following where needed:

```
	&mdio0_pins {
		clk {
			pins = "gpio1";
			input-disable;
		};
	};
```

I'm not sure whether that makes it easier, though.

adrianschmutzler commented:
Another option I'm looking into right now would be to remove the linear inheritence of the DTSI files:

qcom-ipq8064.dtsi
vvv
qcom-ipq8064-v1.0.dtsi
vvv
qcom-ipq8064-v2.0.dtsi
vvv
qcom-ipq8065.dtsi

So, if we would break the interdependence of the lower three (which seems relatively easy), we could define pins for smaller subgroups directly in them (e.g. I think the mdio0_pins difference is ipq8064 vs. ipq8065 ...)

mans0n commented:
> (e.g. I think the mdio0_pins difference is ipq8064 vs. ipq8065 ...)

Oh, I didn't notice that. Good find.

So you mean sourcing ipq8064.dtsi directly in ipq8065.dtsi and defining pinmuxes separately in ipq8064-v1.0.dtsi and ipq8065.dtsi?

adrianschmutzler commented:
> So you mean sourcing ipq8064.dtsi directly in ipq8065.dtsi and defining pinmuxes separately in ipq8064-v1.0.dtsi and ipq8065.dtsi?

I've updated my PR #3164 with a partially different approach and also added a patch for disentangling the DTSI files there (which is conceptionally independent of the other thing).

One could then consider to indeed add globally common definitions in ipq8064.dtsi and some other definitions in the other DTSI files where appropriate.

mans0n commented:
Moved more pinmux nodes.

* uart0_pins: Did not touch.

* i2c4_pins: This node seems to be not used at all in dts.
  Added the most common form in the dtsi. Did not touch rpm_pins.

* spi_pins: The common mux node is already present in the dtsi.
  Removed the duplicate nodes from dts.

* nand_pins: Moved the entire node.

* mdio0_pins: Moved the common mux node.

* rgmii2_pins: Moved the common mux node. "pins" property is overrided in ap161.

So the patch still remains cosmetic, I believe.

adrianschmutzler commented:
Will build-test and merge later today. Thanks for cleaning this up.
